server/src/main/java/org/elasticsearch/repositories/RepositoryData.java

```diff
@@ -615,7 +619,9 @@ public static RepositoryData snapshotsFromXContent(final XContentParser parser,
                         assert indexId != null;
                         indexSnapshots.put(indexId, Collections.unmodifiableList(snapshotIds));
                         for (int i = 0; i < gens.size(); i++) {
-                            shardGenerations.put(indexId, i, gens.get(i));
+                            final String parsedGen = gens.get(i);
+                            shardGenerations.put(
+                                    indexId, i, fixBrokenShardGens ? ShardGenerations.fixShardGeneration(parsedGen) : parsedGen);
```

Comment:
I think we should avoid putting null values into the map (in particular as `ShardGenerations.Builder.put` does not seem to define `@Nullable String generation`, and the semantics of putting null into map can be problematic, as the Map interface says the behavior is underspecified whether a NPE is thrown). I would prefer a boolean method here with conditional logic

Reply:
++ done